RecoLocalCalo/HcalRecAlgos/test/HcalCollapseAnalyzer.cc

```diff
@@ -0,0 +1,173 @@
+  iSetup.get<HcalRecNumberingRecord>().get(htopo);
+  const HcalTopology* theHBHETopology = htopo.product();
+
+  edm::Handle<HBHERecHitCollection> hbhereco;
+  iEvent.getByToken(tok_hbhe_,      hbhereco);
+  edm::Handle<HBHERecHitCollection> hbheprereco;
+  iEvent.getByToken(tok_prehbhe_,   hbheprereco);
+  if (verbosity_ > 0) {
+    edm::LogVerbatim("Collapse") << "Handle Reco "    << hbhereco << " Size " 
+				 << hbhereco->size();
+    edm::LogVerbatim("Collapse") << "Handle PreReco " << hbheprereco <<" Size "
+				 << hbheprereco->size();
+  }
+  if (hbhereco.isValid() && hbheprereco.isValid()) {
+    const HBHERecHitCollection* recohbhe    = hbhereco.product();
+    const HBHERecHitCollection* prerecohbhe = hbheprereco.product();
+    if (verbosity_ > 0) 
+      edm::LogVerbatim("Collapse") << "Size of hbhereco: " << recohbhe->size()
+				   << " and hbheprereco: " << prerecohbhe->size();
+    double sfrac = (prerecohbhe->size() == 0) ? 1 :
+      ((double)(recohbhe->size()))/((double)(prerecohbhe->size()));
+    h_sfrac->Fill(sfrac);
+    h_size->Fill(recohbhe->size());
+    for (const auto & hit : (*recohbhe)) {
+      HcalDetId id = hit.id();
+      if (((id.subdet() == HcalEndcap) && doHE_) ||
+	  ((id.subdet() == HcalBarrel) && doHB_)) {
+	h_depth->Fill(id.depth());
+	std::vector<HcalDetId> ids;
+	theHBHETopology->unmergeDepthDetId(id, ids);
+	if (verbosity_ > 0) {
+	  edm::LogVerbatim("Collapse") << id << " is derived from " << ids.size() 
+				       << " components";
+	  for (unsigned int k=0; k<ids.size(); ++k) 
+	    edm::LogVerbatim("Collapse") << "[" << k << "] " << ids[k];
+	}
+	h_merge->Fill(ids.size());
+	double energy = hit.energy();
+	double etot(0);
+	unsigned int k(0);
+	for (const auto phit : (*prerecohbhe)) {
+	  if (std::find(ids.begin(),ids.end(),phit.id()) != ids.end()) {
+	    etot += phit.energy();
+	    double frac = (energy > 0) ? phit.energy()/energy : 1.0; 
+	    h_frac->Fill(frac);
+	    if (verbosity_ > 0) 
+	      edm::LogVerbatim("Collapse") << "Frac[" << k << "] " << frac 
+					   << " for " <<phit.id();
+	    ++k;
+	  }
+	}
+	double frac = (energy > 0) ? etot/energy : 1.0; 
+        h_balance->Fill(frac);
+	if (verbosity_ > 0) 
+	  edm::LogVerbatim("Collapse") << "Total energy " << energy << ":" << etot
+				       << ":" << frac;
+      }
+    }
+  }
+}
+
+void HcalCollapseAnalyzer::beginJob() {
+  // Book histograms
+  edm::Service<TFileService>             fs;
+  h_merge  = fs->make<TH1I>("h_merge","Number of hits merged",10,0.0,10.0);
+  h_size   = fs->make<TH1I>("h_size","Size of the RecHit collection",100,500.0,1500.0);
+  h_depth  = fs->make<TH1I>("h_depth","Depth of the Id's used",10,0.0,10.0);
+  h_sfrac  = fs->make<TH1D>("h_sfrac","Ratio of sizes of preRecHit and RecHit collections",150,0.0,1.5);
+  h_frac   = fs->make<TH1D>("h_frac", "Fraction of energy before collapse",150,0.0,1.5);
+  h_balance= fs->make<TH1D>("h_balance", "Balance of energy between pre- and post-collapse",100,0.5,1.5);
```

Comment:
This module only fills per-event histograms. This looks like a good case to make it a DQMEDAnalyzer and perhaps even move it to a relevant DQM package.

Reply:
Did not realize it will be reviewed so quickly. OK could be moved